services/src/main/java/com/linkedin/kmf/services/ProduceService.java

```diff
@@ -284,4 +300,47 @@ public void run() {
     }
   }
 
+  /**
+   * This should be periodically run to check for new partitions on the monitored topic.  When new partitions are
+   * detected the executor is shutdown, the producer is shutdown, a new producer is created (the old producer does not
+   * always become aware of the new partitions), new produce runnables are created on a new executor service, new
+   * sensors are added for the new partitions.
+   */
+
+  private class HandleNewPartitions implements Runnable {
+
+    public void run() {
+      int currentPartitionCount = Utils.getPartitionNumForTopic(_zkConnect, _topic);
+      if (currentPartitionCount == _partitionNum.get()) {
+        return;
+      }
+      LOG.info(_name + ": Detected new partitions to monitor.");
+      //TODO: Should the ProduceService exit if we can't restart the producer runnables?
+      _executor.shutdown();
+      try {
+        _executor.awaitTermination(Integer.MAX_VALUE, TimeUnit.MILLISECONDS);
+      } catch (InterruptedException e) {
+        throw new IllegalStateException(e);
+      }
+      _producer.close();
+      int nextPartitionStart = _partitionNum.get();
+      int addedPartitionCount = currentPartitionCount - nextPartitionStart;
+      int newPartitionCount = nextPartitionStart + addedPartitionCount;
+      for (int newPartition = nextPartitionStart; newPartition < newPartitionCount; newPartition++) {
+        LOG.debug(_name + ": Adding " + newPartition + " to sensors.");
```

Comment:
Since we change the partition number and add sensors for new partitions here, we should probably move following logic for add default index for new partitions from scheduleProduceRunnables() to here. What do you think?